flake.nix

```
@@ -0,0 +1,96 @@
+{
+  description = "snarkOS - A decentralized operating system";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    rust-overlay.url = "github:oxalica/rust-overlay";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs =
+    {
+      self,
+      nixpkgs,
+      rust-overlay,
+      flake-utils,
+    }:
+    flake-utils.lib.eachDefaultSystem (
+      system:
+      let
+        overlays = [ (import rust-overlay) ];
+        pkgs = import nixpkgs {
+          inherit system overlays;
+        };
+
+        # Rust toolchain matching rust-toolchain.toml
+        rustToolchain = pkgs.rust-bin.fromRustupToolchainFile ./rust-toolchain.toml;
+
+        # Native build inputs for the project
+        nativeBuildInputs = with pkgs; [
+          rustToolchain
+          curl
+          openssl
+          pkg-config
+          clang
+          gcc
+          lld
+          libclang
+          git # Required for build.rs (built crate with git2 feature)
+        ];
+
+        # Runtime dependencies
+        buildInputs = with pkgs; [
+          openssl
+        ];
+
+        # Read version from Cargo.toml
+        cargoToml = builtins.fromTOML (builtins.readFile ./Cargo.toml);
+        version = cargoToml.package.version;
+      in
+      {
+        devShells.default = pkgs.mkShell {
+          inherit nativeBuildInputs buildInputs;
+
+          # Environment variables
+          LIBCLANG_PATH = "${pkgs.libclang.lib}/lib";
+
+          # Ensure git is available for build.rs
+          GIT = "${pkgs.git}/bin/git";
+
+          shellHook = ''
+            echo "snarkOS development environment"
+            echo "Rust version: $(rustc --version)"
+            echo "Cargo version: $(cargo --version)"
+            echo ""
+            echo "To build the project, run: cargo build --release"
+            echo "To run tests, run: cargo test"
+          '';
```

This is sweet - I don't make much use of shellHook but this a nice use-case!

I wonder we should just keep it informative and minimal?

```suggestion
          shellHook = ''
            echo "snarkOS development environment"
            echo "Rust version: $(rustc --version)"
            echo "Cargo version: $(cargo --version)"
          '';
```